src/TTImp/ProcessDef.idr

```diff
@@ -708,8 +711,17 @@ processDef opts nest env fc n_in cs_in
                        then erased
                        else linear
          nidx <- resolveName n
-         cs <- traverse (checkClause mult (visibility gdef)
+         
+         -- Set the default totality option to the current one, saving the global one
+         defaultTotality <- getDefaultTotalityOption
+         let treq = fromMaybe defaultTotality (findSetTotal (flags gdef))
+         setDefaultTotalityOption treq
+         
+         cs <- traverse (checkClause mult (visibility gdef) treq
                                      hashit nidx opts nest env) cs_in
+         -- Restore global default
+         setDefaultTotalityOption defaultTotality
```

Comment:
A common technique to do this in a clean way is to define a

```idris
withTotality : TotalReq -> Lazy (Core a) -> Core a
withTotality tot c = do
         -- Set the default totality option to the current one, saving the global one
         defaultTotality <- getDefaultTotalityOption
         setDefaultTotalityOption tot
         x <- c
         setDefaultTotalityOption defaultTotality
         pure x
```

and then use that function in your code.

If you want to be totally (ha!) safe, this should probably even wrap the
call to `c` in a `catch` so that you may reset the default even if an error
is thrown (before re-throwing it).

Reply:
Thanks, that's a good suggestion. Errors was a concern for me, and this is an elegant way to do it.
(I thought this would be done more nicely with dynamically bound state + dynamic unwind, but now we're actually doing it like that.)